File: dynamic-programming/longest-decreasing-subsequence/javascript/LDS.js
const nums = [20, 8, 12, 16, 10, 9, 18, 7];

console.log("Length of LDS: " + getLds(nums, 0, Number.MAX_SAFE_INTEGER));

var arr = new Array();

for (let i = 0; i < nums.length + 1; i++) {
    arr.push(new Array(nums.length).fill(-1))
}

console.log("Top Down- Length of LDS: " + getLdsTopDown(nums, -1, 0, arr));

console.log("Bottom up- Length of LDS: " + getLdsBottomUp(nums));

//O(n^2)
function getLdsBottomUp(nums) {
    if (nums.length == 0)
        return 0;

    var maxLds = new Array(nums.length).fill(1);

    let maxSoFar = 1;
    for (let j = 1; j < nums.length; j++) {
        for (let i = 0; i < j; i++) {
            if (nums[j] < nums[i])
                maxLds[j] = Math.max(maxLds[j], maxLds[i] + 1);
        }
        maxSoFar = Math.max(maxSoFar, maxLds[j]);
    }
    return maxSoFar;
}

//O(n^2)
function getLdsTopDown(nums, prevIndex, curr, arr) {

    if (curr == nums.length)
        return 0;

    if (arr[prevIndex + 1][curr] > 0)
        return arr[prevIndex + 1][curr];

    let incl = 0;
    if (prevIndex < 0 || nums[curr] < nums[prevIndex]) {
        incl = 1 + getLdsTopDown(nums, curr, curr + 1, arr);
    }

    let excl = getLdsTopDown(nums, prevIndex, curr + 1, arr);

    //memoization
    arr[prevIndex + 1][curr] = Math.max(incl, excl);

    return arr[prevIndex + 1][curr];
}

//O(2^n)
function getLds(nums, i, prev) {

    if (i == nums.length)
        return 0;

    let incl = 0;
    if (nums[i] < prev)
        incl = 1 + getLds(nums, i + 1, nums[i]);

    let excl = getLds(nums, i + 1, prev);

    return Math.max(incl, excl);
}

File: dynamic-programming/longest-decreasing-subsequence/java/LDS.java
package com.basicsstrong.algorithm.dynamicprogramming;

import java.util.Arrays;

public class LDS {
	static int count=0;

	public static void main(String[] args)
	{
		int[] nums = { 20, 8, 12, 16, 10, 9, 18, 7 };

		System.out.println("Recursive: Length of LDS " + getLDS(nums, 0, Integer.MAX_VALUE));

		int[][] dp = new int[nums.length+1][nums.length];

		for (int[] arr : dp) {
			Arrays.fill(arr, -1);
		}
		System.out.println("DP top down: Length of LDS " + getLDS(nums,-1, 0, dp));

		System.out.println("DP bottom up: Length of LDS " + getLDS(nums));
	}

	
	//Recursive Approach
	public static int getLDS(int[] nums, int i, int prev) {

		if (i == nums.length) {
			return 0;
		}

		int incl = 0;
		if (nums[i] < prev) {
			incl = 1 + getLDS(nums, i + 1, nums[i]);
		}

		int excl = getLDS(nums, i + 1, prev);
  	return Integer.max(incl, excl);
	}

	
	//DP : Top Down Approach
  	public static int getLDS(int[] nums, int prevIndex, int curr, int[][] dp) {
    	if (curr == nums.length) {
			return 0;
		}

		if(dp[prevIndex+1][curr]>0) {
			return dp[prevIndex+1][curr];
		}

	 	int incl = 0;
		if (prevIndex<0 || nums[curr] < nums[prevIndex]) {
			incl = 1 + getLDS(nums,curr,curr+1,dp);
		}

		int excl = getLDS(nums,prevIndex,curr+1,dp);

		dp[prevIndex+1][curr]= Integer.max(incl, excl);
		return dp[prevIndex+1][curr];
	}

  	
  	//DP : Bottom UP Approach
	public static int getLDS(int[] nums) {
		if (nums.length == 0) {
			return 0;
		}

		int[] maxLds = new int[nums.length];
		Arrays.fill(maxLds, 1);

		int maximumSoFar = 1;

		for (int j = 1; j < nums.length; j++) {
			for (int i = 0; i < j; i++) {
				if (nums[j] < nums[i]) {
					maxLds[j] = Math.max(maxLds[j], maxLds[i] + 1);
				}
			}

			maximumSoFar = Math.max(maximumSoFar, maxLds[j]);
		}
		return maximumSoFar;
	}


}
